druntime/src/core/internal/array/capacity.d

```diff
@@ -9,6 +9,181 @@
 */
 module core.internal.array.capacity;
 
+import core.attribute : weak;
+
+// for now, all GC array functions are not exposed via core.memory.
+extern (C)
+{
+    size_t gc_reserveArrayCapacity(void[] slice, size_t request, bool atomic) nothrow pure;
+    bool gc_shrinkArrayUsed(void[] slice, size_t existingUsed, bool atomic) nothrow pure;
+}
+
+
+private uint __typeAttrs(T)(void *copyAttrsFrom = null)
+{
+    import core.internal.traits : hasElaborateDestructor, hasIndirections;
+    import core.memory : GC;
+
+    alias BlkAttr = GC.BlkAttr;
+
+    if (copyAttrsFrom)
+    {
+        // try to copy attrs from the given block
+        auto info = GC.query(copyAttrsFrom);
+        if (info.base)
+            return info.attr;
+    }
+
+    uint attrs = 0;
+    static if (hasIndirections!T)
+        attrs |= BlkAttr.NO_SCAN;
+
+    static if (hasElaborateDestructor!T)
+        attrs |= BlkAttr.FINALIZE;
+
+    return attrs;
+}
+
+private void __doPostblit(T)(T[] arr)
+{
+    // infer static postblit type, run postblit if any
+    static if (__traits(hasPostblit, T))
+    {
+        static if (__traits(isStaticArray, T) && is(T : E[], E))
+            __doPostblit(cast(E[]) arr);
+        else static if (!is(typeof(arr[0].__xpostblit())) && is(immutable T == immutable U, U))
+            foreach (ref elem; (() @trusted => cast(U[]) arr)())
+                elem.__xpostblit();
+        else
+            foreach (ref elem; arr)
+                elem.__xpostblit();
+    }
+}
+
+/**
+Set the array capacity.
+
+If the array capacity isn't currently large enough
+to hold the requested capacity (in number of elements), then the array is
+resized/reallocated to the appropriate size.
+
+Pass in a requested capacity of 0 to get the current capacity.
+
+Params:
+    newcapacity = requested new capacity
+    p = pointer to array to set. Its `length` is left unchanged.
+
+Returns: the number of elements that can actually be stored once the resizing is done
+*/
+size_t _d_arraysetcapacityPureNothrow(T)(size_t newcapacity, void[]* p) pure nothrow @weak @trusted
+do
+{
+    alias PureNothrowType = size_t function(size_t, void[]*) pure nothrow @trusted;
+    return (cast(PureNothrowType) &_d_arraysetcapacity!T)(newcapacity, p);
+}
+
+size_t _d_arraysetcapacity(T)(size_t newcapacity, void[]* p) @weak @trusted
+in
+{
+    assert(!(*p).length || (*p).ptr);
+}
+do
+{
+    import core.exception : onOutOfMemoryError;
+    import core.internal.traits : Unqual;
+    import core.stdc.string : memcpy, memset;
+
+    import core.memory : GC;
+
+    alias BlkAttr = GC.BlkAttr;
+
+    auto isshared = is(T == shared);
+    alias Unqual_T = Unqual!(T);
```

Comment:
You can optimise this: extract `isshared` in `reserve()` and pass it as an argument while instantiating the hook with `Unqual_T` instead. This will reduce the number of `_d_arraysetcapacity` as it would only create a singe one for `shared`, `inout`, `const` qualifiers added to the same type.